Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,775 @@
+  let hp' := ENNReal.toReal_pos_iff.mp hp
+  (if_neg hp'.1.ne').trans (if_neg hp'.2.ne)
+
+theorem prod_edist_eq_sup (f g : WithLp ∞ (α × β)) :
+    edist f g = edist f.fst g.fst ⊔ edist f.snd g.snd := by
+  dsimp [edist]
+  exact if_neg ENNReal.top_ne_zero
+
+end EDist
+
+section EDistProp
+
+variable {α β}
+variable [PseudoEMetricSpace α] [PseudoEMetricSpace β]
+
+/-- This holds independent of `p` and does not require `[Fact (1 ≤ p)]`. We keep it separate
+from `WithLp.instProdPseudoEMetricSpace` so it can be used also for `p < 1`. -/
+theorem prod_edist_self (f : WithLp p (α × β)) : edist f f = 0 := by
+  rcases p.trichotomy with (rfl | rfl | h)
+  · classical
+    simp
+  · simp [prod_edist_eq_sup]
+  · simp [prod_edist_eq_add h, ENNReal.zero_rpow_of_pos h,
+      ENNReal.zero_rpow_of_pos (inv_pos.2 <| h)]
+
+/-- This holds independent of `p` and does not require `[Fact (1 ≤ p)]`. We keep it separate
+from `WithLp.instProdPseudoEMetricSpace` so it can be used also for `p < 1`. -/
+theorem prod_edist_comm (f g : WithLp p (α × β)) : edist f g = edist g f := by
+  classical
+  rcases p.trichotomy with (rfl | rfl | h)
+  · simp [prod_edist_eq_card, eq_comm]
+  · simp only [prod_edist_eq_sup, edist_comm]
+  · simp only [prod_edist_eq_add h, edist_comm]
+
+end EDistProp
+
+section Dist
+
+variable [Dist α] [Dist β]
+
+open Classical in
+/-- Endowing the space `WithLp p (α × β)` with the `L^p` distance. We register this instance
+separate from `WithLp.instProdPseudoMetricSpace` since the latter requires the type class hypothesis
+`[Fact (1 ≤ p)]` in order to prove the triangle inequality.
+
+Registering this separately allows for a future metric-like structure on `WithLp p (α × β)` for
+`p < 1` satisfying a relaxed triangle inequality. The terminology for this varies throughout the
+literature, but it is sometimes called a *quasi-metric* or *semi-metric*. -/
+instance instProdDist : Dist (WithLp p (α × β)) where
+  dist f g :=
+    if _hp : p = 0 then
+      (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1)
+    else if p = ∞ then
+      dist f.fst g.fst ⊔ dist f.snd g.snd
+    else
+      (dist f.fst g.fst ^ p.toReal + dist f.snd g.snd ^ p.toReal) ^ (1 / p.toReal)
+
+variable {p α β}
+
+theorem prod_dist_eq_card [DecidableEq α] [DecidableEq β] (f g : WithLp 0 (α × β)) : dist f g =
+    (if f.fst = g.fst then 0 else 1) + (if f.snd = g.snd then 0 else 1) := by
+  convert if_pos rfl
+
+theorem prod_dist_eq_add (hp : 0 < p.toReal) (f g : WithLp p (α × β)) :
+    dist f g = (dist f.fst g.fst ^ p.toReal + dist f.snd g.snd ^ p.toReal) ^ (1 / p.toReal) :=
+  let hp' := ENNReal.toReal_pos_iff.mp hp
+  (if_neg hp'.1.ne').trans (if_neg hp'.2.ne)
+
+theorem prod_dist_eq_sup (f g : WithLp ∞ (α × β)) :
+    dist f g = dist f.fst g.fst ⊔ dist f.snd g.snd := by
+  dsimp [dist]
+  exact if_neg ENNReal.top_ne_zero
+
+end Dist
+
+section Norm
+
+variable [Norm α] [Zero α] [Norm β] [Zero β]
```

Comment:
I would remove the `Zero` assumptions here, and replace in the definition of the norm the condition `if f.fst = 0` with `if ‖f.fst‖ = 0`.

Reply:
Does this make a difference for pseudo-metric spaces? It seems to me that those definitions aren't equivalent, and it's not clear which one is canonical. I think this spelling matches what we do for `Pi`, and it would be best to br consistent

Reply:
Yes, it does make a difference. In the `p = 0` case, the current definition of the product norm does not depend on the norm on each factor. My point is that it should, instead of depending on another unrelated `Zero` assumption.

Reply:
And we should fix the `PiLp` file in the same way.

Reply:
For instance, if you're taking the product of normed multiplicative groups, the point `0` does not make sense in the group. What makes sense is the point `1`, i.e., the one for which the norm is zero.